### ISSUE TITLE
Implicit initializations of devices at runtime.

### ISSUE THREAD

**Issue by AlexandreEichenberger** [onnx/onnx-mlir#1349]
Right now, the devices are initialized when the dll are loaded, but approach does not work with statically linked libraries.

In a lengthy slack exchange, we derived this solution

Here is the model presented to the user:
* initialization of the devices used for evaluations of graphs are done implicitly for the user. No actions are required. We expect most user to be happy with this simple mode.
* However, if the user wanted to explicitly initialize the devices, we provide a call to do so.
*Further, if a user wanted to shut down the devices, we provide a call to do that as well. It is up to the user to make sure that this call is not concurrent with any graph evaluations. The runtime does not check for this condition and undetermined results occur when this condition is not satisfied.

In more details, here are the steps needed to provide this interface.

1. My understanding is that this “init” must only be called before executing, not creating tensors and or filling them with values. Thus we  propose that inside the run_main_graph, the compiler will automatically add a code of this effect: `if (! OMIsInitializedAccelX) OMInitializeAccelX()`. For 99.999% of the calls, that will corresponds to a single load, compare, and branch per graph evaluation (not per NNPA instruction, for the whole graph). We believe that this should be very reasonably low. Performance runtime such as OpenMP all do that, so this is not novel. The `OMInitializeAccelX()` will be thread safe, so even if multiple graphs are evaluated in parallel, only one thread will initialize, and the other threads that may have concurrently tried to initialize will wait until the “winning” threads complete the initialization.
2. I think it may be good practice to add an explicit `OMInitializeAccelX()` explicit call in our runtime interface in case some systems want to call it explicitly. It is not required, but it might be useful in some cases. Similarly, if “initializing” a device resulted in heavy blocking of resources in an arbitrary device,
3. I think it would be also good practice to add an explicit `OMShutdownAccelX()` . To keep cost low, this call requires that the user ensure that this shutdown is never called while any graph evaluations are in-flights. Results will be undefined when shutdown is called with graph evaluations in flight. The application using onnx-mlir graphs can trivially implement counters of in-flight evaluations if it needed to do so. We don’t want to move this to our side of the runtime as this will requires locks/atomic operations for each graph evaluation, which is too high a price to pay (as it occurs for each call to a graph, not just the first one).
4. Implementation (for us only). Each accelerators being supported in onnx-mlir will provide a (possibly empty) init and shutdown method. When building onnx-mlir for a given set of accelerator S, the `OMInitializeAccelX()` and `OMShutdownAccelX()` will call the init and shutdown for each accelerators in S. Practically, this mean that if the NNPA library has code to initialize the NNPA, we must include a call to this inside the onnx-mlir support for the NNPA accelerator.

So now the code to generate is as follow

1. inside run_onnx_graph() and all named entry points to execute a graph compile for accelerator X: `if (!OMIsInitializedAccelX) OMInitializeAccelX()`.  
    1. we need to add a definition of  OMIsInitializedAccelX set to zero in the OMRuntime .c and .h
    2. we need to add an OMInitializeAccelX() for each supported accelerator in the OMRuntime .c and .h
    3. we need to add the testing code when generating the run_onnx_graph() entry points.
2. We need to add an OMShutdownAccelX() for each access in OMRuntime.c and .h

I need help for 1.3 above.

**Comment by gongsu832** [comment]
I'm not sure what OMRuntimeInitialize()/OMRuntimeShutdown() are supposed to do. The discussions we had on slack and webex were mostly about OMInitializeAccelX()/OMShutdownAccelX().

**Comment by AlexandreEichenberger** [comment]
Updated the original post to reflect the newer names, that should clear the confusion.

**Comment by gongsu832** [comment]
I suggest we focus on the generation of the internal init/fini calls first. Exposing them to the API users should be considered much more carefully since adding a call is easy but removing one will typically be a nightmare. We should only be adding calls if there are clear demands from use cases. I'm not convinced yet why you want to manually do init/fini when they are already done for you automatically.

**Comment by tungld** [comment]
> I need help for 1.3 above.

Do you mean this `we need to add the testing code when generating the run_onnx_graph() entry points.`? If so I can help. I did generate if-then-else for  multiple entry points, it should be the same here for accelerators. The accelerator to test can be obtained from `--maccel`.

**Comment by AlexandreEichenberger** [comment]
> I suggest we focus on the generation of the internal init/fini calls first. 
@gongsu832 The hard part is to generate the internal init/fini calls. They need to be part of the ONNX Runtime.
Exposing the interface is something we can do later

> Do you mean this we need to add the testing code when generating the run_onnx_graph() entry points.? If so I can help.
@tungld yes, that would be great

**Comment by AlexandreEichenberger** [comment]
done with #1357 and another PR for adding calls to this new functions #1374
